### ISSUE TITLE
building the doc

### ISSUE THREAD

**Issue by kiwifb** [cschwan/sage-on-gentoo#210]
It would be nice to have at least an experimental way of building the doc again. Plenty of work has been done upstream that we could try to leverage.

**Comment by strogdon** [comment]
Just noticed today that the sphinx issue has migrated to
https://github.com/sphinx-doc/sphinx/issues/1480

Not that it's important but the pdf-docs do build here with sage-doc-{6.5,9999-r1} with the following caveat. Sage-6.5 introduces some docs in Hungarian. It would seem that ``magyar.ldf`` is needed to complete the build. The dev-texlive/texlive-latex-2012 ebuild installs this file under ``/usr/share/texmf-dist/tex/generic/babel/magyar.ldf``. However with dev-texlive/texlive-latex-2014 the file appears to be missing and one gets strange build errors 
```
! Package babel Error: You haven't specified a language option.
```
But it will build with texlive-latex-2014 if the ``magyar.ldf`` is present and the ``ls-R`` database (mktexlsr) is regenerated.

**Comment by kiwifb** [comment]
Cool we may have live action on sphinx. I must say I rarely try to build the pdf these days. But the language situation is getting ridiculous, have to do some thinking.

**Comment by kiwifb** [comment]
6.6.beta1 is not kind to the ebuild. Will dissect it later.

**Comment by strogdon** [comment]
6.6.beta1 builds here (prefix), html + pdf. But I may have some locally changed dependencies, like sphinx and texlive. Even though it builds it seems to take for ever to install.

**Comment by kiwifb** [comment]
Second time lucky. I will have to keep an eye on that, there may be rare occurrence of failure when building in parallel. The freetype component of matplotlib wanted to open a non existent file.

**Comment by kiwifb** [comment]
http://trac.sagemath.org/ticket/17874

**Comment by strogdon** [comment]
From the Gentoo side the problem , for ``>=texlive-2013``, is that there is no appropriate ``langhungarian``. For ``>=texlive-langhungarian-2013`` there should be ``babel-hungarian`` as part of the ``TEXLIVE_MODULE_CONTENTS`` variable in the ebuild that would install the needed ``magyar.ldf`` file.

**Comment by kiwifb** [comment]
So there is a bug in gentoo about texlive. I am guessing that texlive-langhungarian-201{3,4} would be identical to 2012, is there a dependency problem that prevent using texlive-langhungarian-2012 in texlive-201{3,4}?

**Comment by strogdon** [comment]
I see no problem here in installing texlive-langhungarian-2012 with texlive-201{3,4}. However, langhungarian-2012 does not pull in the ``magyar.ldf`` Language Definition file since the ``TEXLIVE_MODULE_CONTENTS`` is not quite correct.

**Comment by kiwifb** [comment]
OK so this is the bug we have to report in bugzilla about that. I am about to push an updated list of dependencies for sage-doc-9999-r1 based on the sage ticket.

**Comment by kiwifb** [comment]
Should I include a patched ebuild for texlive-langhugarian in the overlay - at least for testing.

**Comment by strogdon** [comment]
Well if texlive-2012, which is stable, is used/required there is no problem. The required Hungarian lang  is provided by ``dev-texlive/texlive-latex-2012`` independent of langhungarian-2012. But if texlive-201{3,4} is used/required then an ebuild will be needed.

**Comment by kiwifb** [comment]
I see. Which means we need better texlive support in the tree. I am guessing we are short on texlive maintainers.

**Comment by kiwifb** [comment]
I just got hit by the "too deeply nested" error
```
LaTeX Warning: Hyper reference `sage/categories/cartesian_product:sage.categori
es.cartesian_product.CartesianProductsCategory' on page 188 undefined on input
line 15301.


! LaTeX Error: Too deeply nested.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...

l.15329 \begin{Verbatim}[commandchars=\\\{\}]/home/extra/portage/sci-mathematics/sage-doc-9999-r1/temp/build.log

?
! Emergency stop.
 ...

l.15329 \begin{Verbatim}[commandchars=\\\{\}]

!  ==> Fatal error occurred, no output PDF file produced!
Transcript written on categories.log.
 [40]Makefile:54: recipe for target 'categories.pdf' failed
make: *** [categories.pdf] Error 1
```
Yet I have the sphinx from the overlay that is supposed to be patched. What's happening to me?

**Comment by kiwifb** [comment]
I think I found where to easily poke and change language built. It may even interest upstream.

**Comment by strogdon** [comment]
Hmm. Just finished building the pdf-docs in prefix here. Check this
```
grep trivlist /usr/lib/python2.7/site-packages/sphinx/texinputs/sphinx.sty
  \trivlist{}{% using trivlist rather than list fixes #777 (too deeply nested error)
  \endtrivlist%
```

**Comment by kiwifb** [comment]
Obviously I don't have it. Very curious. Re-emerging sphinx added the right lines back.

**Comment by kiwifb** [comment]
I have introduced linguas support in 9999-r1. Not sure how it will play with doctesting yet but it does only build a subset of html/pdf :)

I may even submit it for vanilla once we are sure things are working OK. May have to mirror it in sage-doc-bin somehow.

**Comment by strogdon** [comment]
It appears Gentoo texlive ebuilds are using the LINGUAS approach (see https://bugs.gentoo.org/show_bug.cgi?id=542028), although I've not tried it yet to see if it works as described. However, this may complicate things relative sage-doc-9999-r1. Clearly http://trac.sagemath.org/ticket/17874#comment:7 is not approaching things this way.

**Comment by strogdon** [comment]
So I need to fix things here. On the prefix where texlive-2014 stuff is installed I don't have any version of app-text/texlive installed - which could be a problem!

**Comment by strogdon** [comment]
If the ``app-text/texlive`` metapackage ebuild is used then I think ``app-text/texlive[extra,linguas_XX]`` will be the minimum that's needed. There may be some use-flags along the way that the user will have to set to initiate the install. So perhaps something like:
```
@@ -24,7 +24,7 @@
 SRC_URI="${SRC_URI}
 	http://www.mathematik.uni-kl.de/ftp/pub/Math/Singular/SOURCES/3-1-6/Singular-3-1-6-share.tar.gz"
 
-LANGS="ca de en fr hu it pt ru tr"
+LANGS="ca cs de en fr hu it pl pt ru"
 
 LICENSE="GPL-2"
 SLOT="0"
@@ -40,15 +40,17 @@
 DEPEND="!sci-mathematics/sage-doc-bin
 	>=dev-python/docutils-0.12[${PYTHON_USEDEP}]
 	sci-mathematics/sage
-	pdf? ( linguas_ru? ( dev-texlive/texlive-langcyrillic )
-		linguas_fr? ( dev-texlive/texlive-langfrench )
-		linguas_pt? ( dev-texlive/texlive-langportuguese )
-		linguas_de? ( dev-texlive/texlive-langgerman )
-		linguas_hu? ( dev-texlive/texlive-langhungarian )
-		linguas_it? ( dev-texlive/texlive-langitalian )
-		dev-texlive/texlive-langczechslovak
-		linguas_en? ( dev-texlive/texlive-langenglish )
-		dev-texlive/texlive-langpolish )"
+	pdf? ( linguas_ru? ( app-text/texlive[extra,linguas_ru] )
+		linguas_fr? ( app-text/texlive[extra,linguas_fr] )
+		linguas_pt? ( app-text/texlive[extra,linguas_pt] )
+		linguas_de? ( app-text/texlive[extra,linguas_de] )
+		linguas_hu? ( app-text/texlive[extra,linguas_hu] )
+		linguas_it? ( app-text/texlive[extra,linguas_it] )
+		linguas_en? ( app-text/texlive[extra,linguas_en] )
+		linguas_pl? ( app-text/texlive[extra,linguas_pl] )
+		linguas_ca? ( app-text/texlive[extra,linguas_ca] )
+		linguas_cs? ( app-text/texlive[extra,linguas_cs] ) )"
+
 RDEPEND="${DEPEND}
 	>=dev-python/sphinx-1.2.2[${PYTHON_USEDEP}]"
```

**Comment by kiwifb** [comment]
Thanks for the input Steve. I have been quite busy all morning and couldn't look at it myself. I won't add `cs` and `pl` because I am matching what is available under `src/doc`, ` texlive-langczechslovak` and  `texlive-langpolish` must used by another language. I'll have to see what turkish (`tr`) requires.

**Comment by strogdon** [comment]
Yes, ``cs`` and ``pl`` should be deleted. I believe they will be pulled in when needed by the appropriate ``linguas_XX``. And I missed the ``tr``. Here with texlive-2014 the ``linguas_tr`` option pulls in texlive-langeuropean for which there is every language imaginable, well almost every language. And things are slick since, for example, with USE="-linguas_en linguas_fr" just the French docs are built. 

And, additionally, while testing I did encounter your, I believe, parallel build issue:
```
# Sphinx version: 1.2.3
# Python version: 2.7.8
# Docutils version: 0.12 release
# Jinja2 version: 2.7.3
# Loaded extensions:
Traceback (most recent call last):
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sphinx/cmdline.py", line 253, in main
    warningiserror, tags, verbosity, parallel)
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sphinx/application.py", line 119, in __init__
    self.setup_extension(extension)
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/sphinx/application.py", line 333, in setup_extension
    mod = __import__(extension, None, None, ['setup'])
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/sphinxext/plot_directive.py", line 156, in <module>
    import matplotlib.pyplot as plt
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/pyplot.py", line 24, in <module>
    import matplotlib.colorbar
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/colorbar.py", line 29, in <module>
    import matplotlib.collections as collections
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/collections.py", line 23, in <module>
    import matplotlib.backend_bases as backend_bases
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 50, in <module>
    import matplotlib.textpath as textpath
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/textpath.py", line 11, in <module>
    import matplotlib.font_manager as font_manager
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1356, in <module>
    _rebuild()
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/font_manager.py", line 1343, in _rebuild
    pickle_dump(fontManager, _fmcache)
  File "/storage/strogdon/gentoo-redlizard/usr/lib/python2.7/site-packages/matplotlib/font_manager.py", line 939, in pickle_dump
    with open(filename, 'wb') as fh:
IOError: [Errno 2] No such file or directory: '/storage/strogdon/gentoo-redlizard/var/tmp/portage/sci-mathematics/sage-doc-9999-r9/temp/matplotlib-strogdon/fontList.cache'
```

**Comment by kiwifb** [comment]
Yes that's the one I had earlier. It seems there are a couple of concurrency issues in the build.

**Comment by kiwifb** [comment]
Only building specific language will break s few tests. It would be sensible to recommend that all users build `en` as all the pointers default to it.

There are a few more things that can be tightened and I will push experiments on slimming down what's installed further. It will involve a couple of changes in the sage patch set as well. Finally there is one more thing that I'll probably add to the linguas patch to make it more palatable upstream.

I may even consider making building `en` compulsory (if html or pdf is enabled), I think upstream would appreciate it.

Finally, if I don't make building `en` compulsory would building no language at all be acceptable?

**Comment by strogdon** [comment]
So if I have this straight, building no language would imply no docs. Right? So if that's the case then perhaps the user needs to be made aware of that fact. I believe this would affect the help from the notebook.

**Comment by kiwifb** [comment]
Well if I remove the stuff that I think is useless right now it would install singular's help file and some files under common. And yes it would affect the notebook.

Which is why I am thinking of making `en` compulsory. No LINGUAS would be no different than installing with USE="-html -pdf".

**Comment by strogdon** [comment]
So if compulsory then a user could always turn building the docs off. I think I like that, at least relative to the html docs. This is what vanilla does. However, from the vanilla side if one decides to build the 
pdf-docs after the fact then the work of building the html stuff does not have to be redone. Not so with sage-on-gentoo. And here it seems the html stuff is what's time-consuming.

**Comment by kiwifb** [comment]
The ebuild could be split I guess, but originally we didn't want the complexity. Now I almost feel like I will want a virtual to handle sage-doc ;)

**Comment by kiwifb** [comment]
A split ebuild would essentially be in three part:
 * sage-doc-common: stuff in common and singular help file
 * sage-doc-html
 * sage-doc-pdf
No change to sage-doc-bin since we are not building

**Comment by strogdon** [comment]
That sounds like a lot of work.

**Comment by strogdon** [comment]
``sage-doc-9999-r1`` help is now working in the notebook with ``sage-notebook-0.11.4-r1``.

**Comment by kiwifb** [comment]
I think we have things mostly working now. It is time to bid farewell to this issue and open more specific ones as new problem arise or specific task. We have hundreds of comments on various things and it is hardly searchable.
